### PR TITLE
Rework constant tagged string bridging once again in light of new information

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please test

File: stdlib/public/core/StringBridge.swift [hunk 1]
```diff
@@ -18,6 +18,24 @@ internal typealias _CocoaString = AnyObject
 
 #if _runtime(_ObjC)
 
+private var taggedConstantNSStringISAPointer: UInt64 = 0
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift:21]: Computed var to avoid storage (and mutability). Is this really zero?

File: stdlib/public/core/StringBridge.swift [hunk 2]
```diff
@@ -195,6 +239,23 @@ internal func _cocoaStringCopyUTF8(
...
+    )
+    var ptr = mbp.baseAddress._unsafelyUnwrappedUnchecked
+    var numWritten = 0
+    
+    for char in buf {
+      guard numWritten < bufPtr.count else { return nil }
+      // always ascii
+      ptr.initialize(to: char)
+      destination += 1
+      numWritten += 1
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift:255]: How does this loop work? Are we just overwriting the first character repeatedly?

File: stdlib/public/core/StringBridge.swift [hunk 3]
```diff
@@ -239,10 +304,35 @@ private func _NSStringCompare(
...
+//        taggedConstantString.utf16Length
+//      ) == 0
+//    } else {
+//      // can msgSend to them, not tagged constant
+//      return _NSStringCompare(_objc(other), string) == 0
+//    }
+//  }
+  precondition(asFormerlyTaggedConstantString(string) == nil,
+   "Cannot call _cocoaGetCStringTrampoline with a tagged constant NSString"
+  )
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift:332]: What does this have to do with equality?

File: stdlib/public/core/StringBridge.swift [hunk 4]
```diff
@@ -502,17 +577,21 @@ internal func _bridgeCocoaString(_ cocoaString: _CocoaString) -> _StringGuts {
...
-      length: taggedContents.utf16Length
-    )
+    if let taggedContents = untagConstantTaggedCocoaString(cocoaString) {
+      let isaPtr = unsafeBitCast(taggedContents.untaggedCocoa,
+        to: UnsafePointer<_swift_shims_builtin_CFString>
+      )
+      taggedConstantNSStringISAPointer = isaPtr.pointee
+      return _StringGuts(
+        cocoa: taggedContents.untaggedCocoa,
+        providesFastUTF8: false,
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift:588]: Can these give an ASCII pointer?

File: stdlib/public/core/StringBridge.swift [hunk 5]
```diff
@@ -482,11 +561,7 @@ private func getConstantTaggedCocoaContents(_ cocoaString: _CocoaString) ->
   let isASCII = ivarPointer.pointee.flags & isUTF16Mask == 0
   precondition(isASCII) // we don't currently support non-ASCII here
   let contentsPtr = ivarPointer.pointee.str
-  return (
-    utf16Length: Int(length),
-    asciiContentsPointer: contentsPtr,
-    untaggedCocoa: Builtin.reinterpretCast(ivarPointer)
-  )
+  return Builtin.reinterpretCast(ivarPointer)
 #endif
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift:565]: Are ref counting operations allowed on the untagged reference?

File: stdlib/public/core/StringBridge.swift (hunk 4 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift:588]: Yes

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
